jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/util/WebSocketFlusher.java

```diff
@@ -0,0 +1,213 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.websocket.core.util;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.IteratingCallback;
+import org.eclipse.jetty.util.StaticException;
+import org.eclipse.jetty.util.thread.AutoLock;
+import org.eclipse.jetty.websocket.core.OutgoingEntry;
+import org.eclipse.jetty.websocket.core.OutgoingFrames;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is used to iteratively transform or process a frame into one or more other frames.
+ * When a frame is ready to be processed {@link #onFrame(OutgoingEntry, boolean)} is called.
+ * Subsequent calls to {@link #onFrame(OutgoingEntry, boolean)} are after each time the entry is succeeded
+ * until one of these calls returns true to indicate they are done processing the frame and are ready to receive a new one.
+ */
+public abstract class WebSocketFlusher implements OutgoingFrames
+{
+    private static final Throwable SENTINEL_CLOSE_EXCEPTION = new StaticException("Closed");
+    private final Logger log = LoggerFactory.getLogger(this.getClass());
+
+    private final AutoLock _lock = new AutoLock();
+    private final Queue<OutgoingEntry> _entries = new ArrayDeque<>();
+    private final IteratingCallback _flusher = new Flusher();
+    private Throwable _failure;
+
+    /**
+     * Called when a frame is ready to be transformed.
+     * @param entry the entry containing the frame to be transformed.
+     * @param first true if this is the first time this entry is being processed.
+     * @return true to indicate that you have finished transforming this entry.
+     */
+    protected abstract boolean onFrame(OutgoingEntry entry, boolean first);
```

Comment:
Can the entry transformation be asynchronous, so that this should rather be:

```suggestion
    protected abstract void onFrame(OutgoingEntry entry, boolean first, Promise<Boolean> promise);
```

Reply:
I don't think there is a need for this.

And I don't see how this would work with the Flusher.process implementation.